### PR TITLE
Sink binding docs (reopened PR)

### LINKED ISSUE #3510: [1.0] Fully document sinkbindings.sources.knative.dev
@n3wscott commented on [Mon Feb 22 2021](https://github.com/knative/eventing/issues/4918)

Related to https://github.com/knative/docs/issues/3268
The following work needs to be validated for sinkbindings.sources.knative.dev to be ready for 1.0:
- [ ] Getting Started demo using the component.
- [ ] Sample minimum of component
- [ ] Sample complete resource of component
- [ ] Fully documented features and options, including
    - [ ] Explanation of resource.
    - [ ] All spec fields
    - [ ] All status fields
    - [ ] Any magic labels or annotations that apply.
    - [ ] With kubectl cli
    - [ ] With kn cli

---

@snneji commented on [Tue Apr 27 2021](https://github.com/knative/eventing/issues/4918#issuecomment-827616695)

/assign

### PR DESCRIPTION
**Pull request by snneji**
Reopening this PR so that I can apply feedback from @abrennan89.

Closes #3510

Worked with @n3wscott on this so the docs should now provide:

- Getting Started demo using the component.
- Sample minimum of component
- Explanation of resource.
- All spec fields
- With kubectl cli
- With kn cli

### CODE REVIEW

**Comment by abrennan89** [comment]
@thisisnotapril could you please fix the CLA thing? 🙏🏻 

@snneji there seem to be a lot of tests failing on this one besides the CLA?

File: docs/eventing/sources/sinkbinding/README.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+---
+title: "SinkBinding"
+weight: 01
+type: "docs"
+aliases:
+  - /docs/eventing/samples/sinkbinding/
+  - /docs/eventing/sources/sinkbinding
+---
+
```
Comment (abrennan89) [inline docs/eventing/sources/sinkbinding/README.md:9]: I'd go straight for removing these and adding the redirects etc to the mkdocs.yml, and adding a H1 here instead

File: docs/eventing/sources/sinkbinding/getting-started.md [hunk 2]
```diff
@@ -0,0 +1,300 @@
+---
+title: "Create a SinkBinding object"
+weight: 02
+type: "docs"
+---
+
```
Comment (abrennan89) [inline docs/eventing/sources/sinkbinding/getting-started.md:6]: same for this

File: docs/eventing/sources/sinkbinding/getting-started.md [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+        For example, `gcr.io/knative-releases/knative.dev/eventing-contrib/cmd/event_display`
+   </br></br>
+    2. Apply the YAML file by running:
+        ```bash
+        kubectl apply --filename service.yaml
+      ```
+
+
+
+
```
Comment (abrennan89) [inline docs/eventing/sources/sinkbinding/getting-started.md:111]: remove unnecessary whitespace?

File: docs/eventing/sources/sinkbinding/getting-started.md [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+        ```
+        Where:
+        - `<app-name>` is the name of the application. For example, `event-display`.
+        - `<image-url>` is the URL of the image container.
+        For example, `gcr.io/knative-releases/knative.dev/eventing-contrib/cmd/event_display`
+   </br></br>
+    2. Apply the YAML file by running:
+        ```bash
+        kubectl apply --filename service.yaml
+      ```
```
Comment (abrennan89) [inline docs/eventing/sources/sinkbinding/getting-started.md:107]: per style guide, make this one step with a command that's
```yaml
kubectl apply -f - <<EOF
# code lines
EOF
```

make this consistent for all creating yaml bits

**Review by abrennan89 — CHANGES_REQUESTED** [review]
EDIT: Just realized you reopened this to apply feedback, so I won't comment on everything til you're done 😉 

Left comments, also `Sinkbinding` vs SinkBinding is still inconsistently formatted, I'd choose one way and go with it.

**Comment by snneji** [comment]
Hi @n3wscott and @abrennan89, would you be able to comment @ googlebot I consent.

**Comment by snneji** [comment]
@abrennan89, I think I addressed the comments that you left on #3648. I was unable to apply edits on that PR, sorry for the confusion.

File: docs/eventing/sources/sinkbinding/getting-started.md [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+If you have an existing subject and sink, you can replace the examples with your
+own values.
+
+## Before you begin
+
+Before you can create a SinkBinding object:
+
+- You must have Knative Eventing installed on your cluster.
+- Optional: If you want to use `kn` commands with SinkBinding, install the `kn` CLI.
+
```
Comment (abrennan89) [inline docs/eventing/sources/sinkbinding/getting-started.md:20]: I'd add links for the install docs

File: docs/eventing/sources/sinkbinding/getting-started.md [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+The SinkBinding object operates in one of two modes: `exclusion` or `inclusion`.
+
+The default mode is `exclusion`.
+In exclusion mode, SinkBinding behavior is enabled for the namespace by default.
+To disallow a namespace from being evaluated for mutation you must exclude it
+using the label `bindings.knative.dev/exclude: true`.
+
+In inclusion mode, SinkBinding behavior is not enabled for the namespace.
+Before a namespace can be evaluated for mutation, you must
+explicitly include it using the label `bindings.knative.dev/include: true`.
```
Comment (abrennan89) [inline docs/eventing/sources/sinkbinding/getting-started.md:33]: It seems weird that lines are breaking throughout in this PR, like I'm not sure why this line ends here at "must"? Not a big deal, just weird.

File: docs/eventing/sources/sinkbinding/getting-started.md [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+              containers:
+                - image: <image-url>
+        EOF
+        ```
+        Where:
+        - `<app-name>` is the name of the application. For example, `event-display`.
+        - `<image-url>` is the URL of the image container.
+        For example, `gcr.io/knative-releases/knative.dev/eventing-contrib/cmd/event_display`
+
+
```
Comment (abrennan89) [inline docs/eventing/sources/sinkbinding/getting-started.md:114]: @omerbensaadon this is an example of where we need content reuse. As I mentioned in [Slack](https://knative.slack.com/archives/C9CV04DNJ/p1622065546110800) there are waaaaaay too many duplicated docs sections explaining the basics of creating a basic service. We should have one good reusable example that we turn into a snippet.

File: docs/eventing/sources/sinkbinding/reference.md [hunk 8]
```diff
@@ -0,0 +1,171 @@
+# SinkBinding Reference
+
+This topic provides reference information about the configurable fields for the
+SinkBinding object.
+
```
Comment (abrennan89) [inline docs/eventing/sources/sinkbinding/reference.md:5]: extra empty line

File: docs/eventing/sources/sinkbinding/reference.md [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+| Field | Description | Required or optional |
+|-------|-------------|----------------------|
+| [`apiVersion`][kubernetes-overview] | Specifies the API version, for example `sources.knative.dev/v1`. | Required |
+| [`kind`][kubernetes-overview] | Identifies this resource object as a `SinkBinding` object. | Required |
+| [`metadata`][kubernetes-overview] | Specifies metadata that uniquely identifies the `SinkBinding` object. For example, a `name`. | Required |
+| [`spec`][kubernetes-overview] | Specifies the configuration information for this `SinkBinding` object. | Required |
+| [`spec.sink`](#sink-parameter) | A reference to an object that resolves to a URI to use as the sink. | Required |
+| [`spec.subject`](#subject-parameter) | A reference to the resources for which the "runtime contract" is augmented by Binding implementations. | Required |
+| [`spec.ceOverrides`](#cloudevent-overrides) | Defines overrides to control the output format and modifications to the event sent to the sink. | Optional |
+
```
Comment (abrennan89) [inline docs/eventing/sources/sinkbinding/reference.md:20]: again, extra unnecessary line space - maybe adjust whitespace settings in your text editor?

File: docs/eventing/sources/sinkbinding/reference.md [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+    apiVersion: v1
+    kind: Service
+    namespace: default
+    name: mysink
+  uri: /extra/path
+```
+
+**Contract:** This results in the `K_SINK` environment variable being set on the
+`subject` as `"http://mysink.default.svc.cluster.local/extra/path"`.
+
```
Comment (abrennan89) [inline docs/eventing/sources/sinkbinding/reference.md:60]: line space

File: docs/eventing/sources/sinkbinding/reference.md [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+    namespace: default
+    selector:
+      - matchExpression:
+        key: working
+        operator: In
+        values:
+          - example
+          - sample
+  ```
+
```
Comment (abrennan89) [inline docs/eventing/sources/sinkbinding/reference.md:128]: line space

**Review by abrennan89 — CHANGES_REQUESTED** [review]
few mostly minor comments, watch the empty whitespace (double blank lines in a few places), otherwise looks fine

**Comment by abrennan89** [comment]
@googlebot I consent.

**Comment by abrennan89** [comment]
/approve

I'll let @n3wscott or someone lgtm this but looks OK to me

**Comment by thisisnotapril** [comment]
@abrennan89 bot happy!

**Comment by n3wscott** [comment]
/lgtm

Woooo

/approve

File: docs/eventing/sources/sinkbinding/getting-started.md (hunk 6 above)
Comment (snneji) [reply docs/eventing/sources/sinkbinding/getting-started.md:33]: My text editor breaks lines at the 80 character mark to minimize horizontal scrolling. I probably removed a few words from that line as well. Line breaks such as these will not show up on the site.

File: docs/eventing/sources/sinkbinding/reference.md (hunk 9 above)
Comment (snneji) [reply docs/eventing/sources/sinkbinding/reference.md:20]: I add extra lines around sections to make it easier for me to read a long topic in a text editor. 

Extra lines have never caused problems in the past, but if they cause problems in the Knative docs let me know and I can add a reminder to remove them when I finish a topic.

Comment (abrennan89) [reply docs/eventing/sources/sinkbinding/reference.md:20]: It's really more for consistency, I would really prefer we don't have trailing whitespace for no reason. We used to have a check on github that would fail a test if we had this, but I guess it was removed.

@omerbensaadon it feels like we should have some guidelines on this to keep the docs consistent, e.g. so we don't have double spaces in some places but not others and folks adding new docs know how we expect them to be formatted. I'll leave it up to ya'll to decide which to use since I've already stated my preference, but I think we should at least be consistent.
